nrf-softdevice/src/ble/advertisement_builder.rs

```diff
@@ -0,0 +1,268 @@
+    BroadcastAudioAnnouncement,
+    CommonAudio,
+    HearingAccess,
+    TelephonyAndMediaAudio,
+    PublicBroadcastAnnouncement,
+    ElectronicShelfLabel,
+    GamingAudio,
+    MeshProxySolicitation,
+}
+
+impl Service for BasicService {
+    const SIZE: usize = 2;
+
+    fn render(self, adv: &mut AdvertisementData) {
+        let data = (self as u16).swap_bytes().to_be_bytes();
+        adv.write(&data);
+    }
+}
+
+pub struct CustomService(pub [u8; 16]);
+
+impl Service for CustomService {
+    const SIZE: usize = 16;
+
+    fn render(mut self, adv: &mut AdvertisementData) {
+        self.0.reverse();
+        adv.write(&self.0);
+    }
+}
+
+pub struct Incomplete16<const N: usize>(pub [BasicService; N]);
+pub struct Complete16<const N: usize>(pub [BasicService; N]);
+pub struct Incomplete128<const N: usize>(pub [CustomService; N]);
+pub struct Complete128<const N: usize>(pub [CustomService; N]);
+
+macro_rules! impl_service_list {
+    ($LIST:ident, $SERVICE:ident, $AD:ident) => {
+        impl<const N: usize> ServiceList<$SERVICE, N> for $LIST<N> {
+            const AD: ADType = ADType::$AD;
+
+            fn list(self) -> [$SERVICE; N] {
+                self.0
+            }
+        }
+    };
+}
+
+impl_service_list!(Incomplete16, BasicService, Incomplete16ServiceList);
+impl_service_list!(Complete16, BasicService, Complete16ServiceList);
+impl_service_list!(Incomplete128, CustomService, Incomplete128ServiceList);
+impl_service_list!(Complete128, CustomService, Complete128ServiceList);
+
+pub trait Name {
+    const AD: ADType;
+
+    fn inner(&self) -> &str;
+}
+
+pub struct ShortName<'a>(pub &'a str);
+pub struct FullName<'a>(pub &'a str);
+
+macro_rules! impl_name {
+    ($NAME:ident, $AD:ident) => {
+        impl<'a> Name for $NAME<'a> {
+            const AD: ADType = ADType::$AD;
+
+            fn inner(&self) -> &str {
+                self.0
+            }
+        }
+    };
+}
+
+impl_name!(ShortName, ShortName);
+impl_name!(FullName, FullName);
+
+pub struct AdvertisementData {
+    buf: [u8; ADV_LEN],
+    ptr: usize,
+}
+
+impl AdvertisementData {
+    pub fn new() -> Self {
+        Self {
+            buf: [0; ADV_LEN],
+            ptr: 0,
+        }
+    }
+
+    fn write(&mut self, data: &[u8]) {
+        let end = self.ptr + data.len();
+
+        if end < ADV_LEN {
+            self.buf[self.ptr..end].copy_from_slice(data);
+        }
+
+        self.ptr += data.len();
+    }
+
+    /// Write raw bytes to the advertisement data.
+    ///
+    /// *Note: The length is automatically computed and prepended.*
+    pub fn raw(mut self, ad: ADType, data: &[u8]) -> Self {
+        self.write(&[data.len() as u8 + 1, ad as u8]);
+        self.write(data);
+
+        self
+    }
+
+    /// View the resulting advertisement data in the form of a byte slice.
+    pub fn as_slice(&self) -> Result<&[u8], Error> {
+        (self.ptr < ADV_LEN)
+            .then(|| &self.buf[..self.ptr])
+            .ok_or(Error::Oversize)
```

Comment:
Since you've tracked how much data was attempted to be written, it would be nice for the `Oversize` error to indicate what size buffer would be required.